cmd/onboard.go

```diff
@@ -226,49 +238,104 @@ TO1:
-			host := *to2Addr.DNSAddress
-			to2URLs = append(to2URLs, scheme+net.JoinHostPort(host, port))
+		if to2Addr.DNSAddress != nil {
+			if isResolvableDNS(*to2Addr.DNSAddress) {
+				host := *to2Addr.DNSAddress
+				ownerURLs = append(ownerURLs, scheme+net.JoinHostPort(host, port))
+			} else {
+				slog.Warn("DNS address is not resolvable", "dns", *to2Addr.DNSAddress)
+			}
 		}
 
 		// Check and add IP address if valid
-		if to2Addr.IPAddress != nil && isValidIP(to2Addr.IPAddress.String()) {
-			host := to2Addr.IPAddress.String()
-			to2URLs = append(to2URLs, scheme+net.JoinHostPort(host, port))
+		if to2Addr.IPAddress != nil {
+			if isValidIP(to2Addr.IPAddress.String()) {
+				host := to2Addr.IPAddress.String()
+				ownerURLs = append(ownerURLs, scheme+net.JoinHostPort(host, port))
+			} else {
+				slog.Warn("IP address is not valid", "ip", to2Addr.IPAddress.String())
+			}
 		}
 	}
 
-	if len(to2URLs) == 0 {
-		return nil, errors.Join(to2URLsErrors...)
+	// Check if TO1 succeeded but returned no valid TO2 addresses
+	if len(ownerURLs) == 0 {
+		slog.Error("TO1 succeeded but no valid TO2 addresses found")
 	}
 
-	// Try TO2 on each address only once
-	var (
-		errs  []error
-		newDC *fdo.DeviceCredential
-	)
-	for _, baseURL := range to2URLs {
-		var err error
-		newDC, err = transferOwnership2(tls.TlsTransport(baseURL, nil, insecureTLS), to1d, conf)
-		if newDC != nil {
-			return newDC, nil
-		}
-		errs = append(errs, err)
+	return ownerURLs, to1d
+}
+
+func transferOwnership(ctx context.Context, rvInfo [][]protocol.RvInstruction, conf fdo.TO2Config) (*fdo.DeviceCredential, error) { //nolint:gocyclo
+	directives := protocol.ParseDeviceRvInfo(rvInfo)
+
+	if len(directives) == 0 {
+		return nil, errors.New("no rendezvous information found that's usable for the device")
 	}
 
-	return nil, errors.Join(errs...)
+	// Infinite retry loop - continues until onboarding succeeds or context canceled
+	for {
+		for i, directive := range directives {
+			isLastDirective := (i == len(directives)-1)
+
+			// Step 1: Get Owner URLs (via TO1 or RV bypass)
+			ownerURLs, to1d := getOwnerURLs(ctx, &directive, conf)
+
+			// Step 2: Attempt TO2 with each Owner URL
+			// Note: If TO1 failed, ownerURLs is empty and loop is skipped
+			if len(ownerURLs) > 0 {
+				slog.Info("Attempting TO2 protocol")
+			}
+			for j, baseURL := range ownerURLs {
+				isLastURL := (j == len(ownerURLs)-1)
+				newDC, err := transferOwnership2(ctx, tls.TlsTransport(baseURL, nil, insecureTLS), to1d, conf)
+				if newDC != nil {
+					slog.Info("TO2 succeeded", "base URL", baseURL)
+					return newDC, nil
+				}
+				slog.Error("TO2 failed", "base URL", baseURL, "error", err)
+
+				// Apply ~120s delay after TO2 failure (last URL skips to Step 3)
+				if !isLastURL {
+					delay := addJitter(120 * time.Second)
+					slog.Info("Applying delay after TO2 failure", "delay", delay)
+					if err := applyDelay(ctx, delay); err != nil {
+						return nil, err
+					}
+				}
```

Comment:
This is a delay _within_ a single directive. The FDO spec only speaks to delays _between_ directives.

There's no reason we can't introduce this delay, but it goes beyond spec and therefore I believe it should be configurable via CLI flags, at the least.

Reply:
Agreed!
Added a new config flag `--to2-retry-delay` default to 0 (i.e. disabled)
keeping default value to 0sec  makes it spec-compliant behavior (no delay between Owner URLs). If user is  experiencing "same-server hammering" issue can opt-in with --to2-retry-delay configurable to desired value. This also helps fast fail-over for the common case while allowing flexibility when needed.
How does that look?
@ben-krieger @kgiusti @runcom @mmartinv